Redis Cache: support more complex types

Ladicek opened the pull request:
This includes generic types, like `List<String>`, array types, like `int[]`, and more.

The type parser in this commit is reasonably complete. Omitting type variables should not be an issue; nested types might, but adding support for them should be doable later.

Fixes #41301

Ladicek committed:
Redis Cache: support more complex types

This includes generic types, like `List<String>`, array types,
like `int[]`, and more.

The type parser in this commit is reasonably complete. Omitting
type variables should not be an issue; nested types might, but
adding support for them should be doable later.

Ladicek commented:
Draft because there's one implementation TODO that I need to think about -- but the rest should be ready for review.

I'm parsing types at runtime and not at build time, because transferring types through the recorder mechanism is not exactly possible. I could easily copy the ArC code that turns `org.jboss.jandex.Type` into `java.lang.reflect.Type`, but that is fundamentally incompatible with the recorder mechanism and I don't see a way to circumvent that.

cescoffier reviewed (commented):
The parser is pretty neat!

Ladicek commented:
2 things I just thought about:

It should be possible to construct the `java.lang.reflect.Type` at build time (because loading `Class`es is possible, and the rest is just a few interfaces that we can implement on our own, as demonstrated in this PR), and that should be straightforward to transfer to runtime via a recorder. Not sure if that would be a huge improvement, but a build-time failure is always preferable to a runtime failure, so I guess I'll work on that.

Maybe the API shouldn't accept `Type`, but `jakarta.enterprise.util.TypeLiteral` (or `com.fasterxml.jackson.core.type.TypeReference`, but I guess I'd prefer the CDI class, as it's more neutral). The implementation would still work on `Type`s, but that would be hidden. Sounds like a small but neat improvement, I'll see how that works.

Ladicek commented:
I think this is ready now. I've implemented both ideas from above, so the type parser is now in `core/deployment` (and `public`). I've also added some tests for it.

To the methods on `RedisCache` that accept `Class`, I added a variant that accepts `TypeLiteral`. This meant that the `Class`-less variants (inherited from `Cache`) can no longer be `default`; I had to move their implementation into `RedisCacheImpl` (which should be OK). This led me to question the usefulness of `RedisCache.getDefaultValueType()` -- I didn't find a use for that, except for implementing the `default` methods, so I marked that method as `@Deprecated`.

Ladicek commented:
One more thing that perhaps needs @cescoffier's attention: I added a TODO to the implementation of `getOrNull()`, because that method accepts a type of the value, but doesn't use it. Instead, it always uses the default type of value. Seems wrong to me, but there's no documentation for the method, so who knows :shrug:

cescoffier commented on extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCacheImpl.java:
```diff
@@ -296,28 +311,36 @@ private void enforceDefaultType() {
             public Uni<V> apply(RedisConnection redisConnection) {
-                return (Uni<V>) doGet(redisConnection, encodedKey, classOfValue, marshaller);
+                // TODO maybe use `type` (if non-null?) instead of `classOfValue`?
+                return doGet(redisConnection, encodedKey, classOfValue, marshaller);
```
Yes, that's a bug 😅

Ladicek replied:
I have another commit that makes the `RedisCache` API overall more regular (all `get*` methods have 3 variants: without type, using the default value type, with `Class` type, and with `TypeLiteral` type). I'll submit that in another PR so we can discuss properly, if you agree.

Ladicek commented:
Thank you!
